### PR TITLE
test for align watermark flow with port configuration

### PR DESCRIPTION
**Pull request by dbarashinvd**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### What is the motivation for this PR?
add a verification test for feature align watermark flow with port configuration

#### How did you do it?
new verification test

#### How did you verify/test it?
run it on several switches

#### Any platform specific information?
not platform specific

#### Supported testbed topology if it's a new test case?
any topology

### CODE REVIEW

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        ConterpollHelper.disable_counterpoll(duthost, CounterpollConstants.COUNTERPOLL_MAPPING.values())
+
+    # verify relevant counterpolls (queue/watermark/pg-drop) are disabled
+    with allure.step("Verifying initial output of {} on {} ..."
+                     .format(CounterpollConstants.COUNTERPOLL_SHOW, duthost.hostname)):
+        verify_all_counterpoll_status(duthost, DISABLE)
+
+    with allure.step("saving config on dut {} after counterpoll disable...".format(duthost.hostname)):
+        duthost.shell('config save -y')
+
```
Comment (bingwang-ms) [inline tests/platform_tests/counterpoll/test_counterpoll_watermark.py:96]: Please consider backup the existing config_db.json before running config save. We have a fixture `backup_and_restore_config_db` to do that

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            reboot(duthost, localhost)
+    # verify all counterpolls are disabled after reload or reboot
+    with allure.step("Verifying output of {} on {} after {} ..."
+                     .format(CounterpollConstants.COUNTERPOLL_SHOW, duthost.hostname, config_apply_method)):
+        verify_all_counterpoll_status(duthost, DISABLE)
+    # enable the selected counterpoll queue/watermark/pg-drop
+    with allure.step("enabling randomly selected counterpoll {} on {} ..."
+                     .format(duthost.hostname, [tested_counterpoll])):
+        ConterpollHelper.enable_counterpoll(duthost, [tested_counterpoll])
+    verify_counterpoll_status(duthost, [tested_counterpoll], ENABLE)
```
Comment (bingwang-ms) [inline tests/platform_tests/counterpoll/test_counterpoll_watermark.py:113]: Align this line into the with block?

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+                    counted += 1
+            for unexpected in unexpected_types:
+                if unexpected in line:
+                    failed_list.append("found for {} unexpected stat counter in FLEX_COUNTER_DB: {}"
+                                       .format(tested_counterpoll, line))
+        logging.info("counted {} {} STATs type in FLEX_COUNTER_DB on {}..."
+                     .format(counted, tested_counterpoll, duthost.hostname))
+        failed_list
+        pytest_assert(counted > 0, "counted {} for {}".format(counted, tested_counterpoll))
+
```
Comment (bingwang-ms) [inline tests/platform_tests/counterpoll/test_counterpoll_watermark.py:167]: Do we need assert `unexpected == 0` as well?

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        with allure.step("counting {} STATS in FLEX_COUNTER_DB on {}...".format(tested_counterpoll, duthost.hostname)):
+            count_watermark_stats_in_counters_db(duthost)
+
+    pytest_assert([] == failed_list, failed_list)
+
+    # no need for reload or reboot when enabling all queue/watermark/pg-drop counterpolls
+    with allure.step("enable all {} counterpolls on {} and check results..."
+                     .format(RELEVANT_COUNTERPOLLS, duthost.hostname)):
+        ConterpollHelper.enable_counterpoll(duthost, RELEVANT_COUNTERPOLLS)
+    verify_counterpoll_status(duthost, RELEVANT_COUNTERPOLLS, ENABLE)
```
Comment (bingwang-ms) [inline tests/platform_tests/counterpoll/test_counterpoll_watermark.py:179]: Same as above. Should `verify_counterpoll_status(duthost, RELEVANT_COUNTERPOLLS, ENABLE)` in the with block?

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 1 above)
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py:96]: isn't it being done automatically when running the test ?
I noticed that the counterpoll configuration was saved and restored when I ran partial test during the development before enabling all counterpolls part at the end of this test.
i.e. when my test only wanted to test one of the counterpoll enabled whlie all of the others disabled, and the switch had all counterpoll enabled, this state of all counterpolls enabled was restored at the end of the test

if needed I can add this fixture usage

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 2 above)
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py:113]: verify_counterpoll_status isn't part of the enablement but can be aligned under this with block and change the comment printed accordingly

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 4 above)
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py:179]: same here, verify_counterpoll_status isn't part of the enablement but can be aligned under this with block and change the comment printed accordingly

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 3 above)
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py:167]: it should have been handled by the failed_list instead, will fix it and update the PR

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            for expected in expected_types:
+                if expected in line:
+                    counted += 1
+            for unexpected in unexpected_types:
+                if unexpected in line:
+                    failed_list.append("found for {} unexpected stat counter in FLEX_COUNTER_DB: {}"
+                                       .format(tested_counterpoll, line))
+        logging.info("counted {} {} STATs type in FLEX_COUNTER_DB on {}..."
+                     .format(counted, tested_counterpoll, duthost.hostname))
+        failed_list
```
[the remark this answers is not in the document]
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py:165]: should be asserted to make sure we don't find other counterpolls when they are not enabled, as it was before the implementation PR was introduced

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 2 above)
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py]: Done

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 4 above)
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py]: Done

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 3 above)
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py:167]: Added assertion for failed_list, it's accumulating the unexpected if there are any

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py]: Done

**Review by wsycqyz — APPROVED** [review]

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 1 above)
Comment (dbarashinvd) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py:99]: fixed, added the use of backup_and_restore_config_db fixture:

17:03:52 duthost_utils._backup_and_restore_config L0038 INFO   | Restore /etc/sonic/config_db.json with /etc/sonic/config_db.json.before_test_function

**Review by bingwang-ms — APPROVED** [review]

**Pull request merged**